### PR TITLE
fix(aws-lambda): make CloudFrontHeaders values optional

### PR DESCRIPTION
**Pull request by bradleyayers**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

It's impossible for every header to be supplied in a request. This PR expresses that in the `CloudFrontHeaders` interface.

### CODE REVIEW

**Comment by simonbuchan** [comment]
The description of the problem here is technically off - the current type only requires whatever values are present to, I think you are hitting a misfeature of TS where optional properties (with `?`) have `undefined` added to their type even when they will never be present with that value.
You can simply workaround this by adding `!` to the value, but it's not great.
That said this workaround should be OK, even if it's a bit ugly, but I'll hold off on at explicit review for someone else to chip in, since it's something that affects all mapping types.
Also if you could add a type test demonstrating the problem that would be good.

**Comment by bradleyayers** [comment]
Sorry I did a really terrible job of describing this issue. Let me try again: the scenario I have is this shape:

```ts
export const handler: AWSLambda.CloudFrontRequestHandler = function handler(
  event,
  context,
  callback
) {
  const { request } = event.Records[0].cf;

  if (request.headers.xfoobar[0].value === 'baz') {
    // do something
  }

  callback(null, request);
};
```

There's no safety around checking that `xfoobar` exists on `request.headers`, so the index into it can throw an exception. I think it's desirable to have to verify that a header exists first before indexing into its array of records. 

Even better would be declaring some headers as always-present (e.g. `host`), while leaving the rest as optional. So perhaps something like this:

```ts

export interface CloudFrontHeadersItem {
  key: string;
  value: string;
}

/**
 * CloudFront events
 * http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-event-structure.html
 */
export interface CloudFrontHeaders {
  host: CloudFrontHeadersItem;
  [name: string]: Array<CloudFrontHeadersItem> | undefined;
}
```

In this case the headers probably need to be split out

**Comment by weswigham** [comment]
See https://github.com/Microsoft/TypeScript/issues/13778
